### PR TITLE
ZCS-5104 - Build controls for set recovery email flow in Preferences Accounts Page

### PR DESCRIPTION
**Pull request by naijenn**
The flow is built and is working with the API methods. 

Some attributes are showing up as null, but these attributes are more for controlling reset and resend limitations. I added TODOs in the doc blocks to mark potential methods to help incorporate these attributes once they are working. 

This will require zm-mailbox feature/PasswordReset branch to see the work in place. Login work will continue on ZCS-5103.

### CODE REVIEW

**Comment by naijenn** [comment]
Test notes have been added to the ticket ZCS-5104.

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 1]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+ * Response object will have a code and msg value to work with.
+ * Determine message to display and update the error message div.
+ *
+ * @param {object} response The response object.
+ */
+ZmAccountsPage.prototype._recoveryErrorCallback =
+function(response) {
+	var message = response.msg;
+	var code = response.code;
+	var messageBeginning = message.indexOf(": ") + 2; // Simplify the message to description - usually the first colon is the error prefix
```
Comment (silentsakky) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: are we getting error message which is translated in user's language?

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 2]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+ *
+ * @TODO Possibly remove this.recoveryEmailStatus if variable is no longer being used in multiple methods
+ * @param {string} status this.RECOVERY_EMAIL_STATUS (none || pending || verified)
+ */
+ZmAccountsPage.prototype._setAccountPasswordControls =
+function(status) {
+	var userStatusMessage;
+	this.recoveryEmailStatus = status;
+	switch(status) {
+		case this.RECOVERY_EMAIL_STATUS.verified: {
```
Comment (silentsakky) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: you don't need curly braces surrounding case statements

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 3]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+
+/*
+ * Add Email Input's Keyup Listener.
+ *
+ * Keyup handler to enable the add button once the input is considered valid.
+ *
+ */
+ZmAccountsPage.prototype._handleRecoveryEmailInput =
+function(result) {
+	var patt = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (silentsakky) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: better to use AjxEmailAddress.isValid for email address validation

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 4]
```diff
@@ -318,6 +318,381 @@ function(params) {
 	this._oneTimeCodesDialog.popup();
 };
 
+/*
+ * Recovery Account Section
+ */
+
+/*
```
Comment (desouzas) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: `/*` style docblock is intentional?

File: WebRoot/templates/prefs/Pages.template [hunk 5]
```diff
@@ -1005,7 +1005,7 @@ and the container element is replaced with that control.
 						<td class='ZOptionsField' valign='middle'><$=ZmMsg.manageSignatures$></td>
 					</tr>
 					<tr><td colspan=2><hr></td></tr>
-					<$ if (appCtxt.get(ZmSetting.TWO_FACTOR_AUTH_AVAILABLE)) { $>
+					<$ // if (appCtxt.get(ZmSetting.TWO_FACTOR_AUTH_AVAILABLE)) { $>
```
Comment (silentsakky) [inline WebRoot/templates/prefs/Pages.template]: looks like commited by mistake

File: WebRoot/templates/prefs/Pages.template [hunk 6]
```diff
@@ -1087,6 +1087,77 @@ and the container element is replaced with that control.
...
+							<td style='text-align:left' colspan=2>
+								<b><$=ZmMsg.recoveryEmailPrefTitle$></b>
+							</td>
+						</tr>
+						<tr>
+							<td></td>
+							<td>
+								<table role="presentation" width=100% class='ZPropertySheet' cellspacing='6'>
+									<tr>
+										<td class='DwtListView-Column'><div style='padding-left:6px; padding-right:6px; color: #727272'>Account</div></td>
```
Comment (silentsakky) [inline WebRoot/templates/prefs/Pages.template]: All strings should be translated

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 7]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+};
+
+/*
+ * Reset Button's Listener.
+ *
+ * Reset code and email input through the SetRecoveryAccountRequest API Method.
+ *
+ */
+ZmAccountsPage.prototype._handleResetRecoveryEmailButton =
+	function() {
```
Comment (desouzas) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Why is this on a new line, but the `appCtxt....` part is beyond what would be a similar line length limit?

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 8]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+ *
+ * Expecting a blank response so the callback acts as confirmation that the code and email has been reset.
+ * Clear the code and email input fields, reset the buttons, update controls based on "null" status.
+ *
+ * @TODO use this.recoveryCodeValidity to manage code verification
+ * @TODO use this.recoveryResetExpiry to enable / disable reset button - might need a timer?
+ * @param {object} response The response object.
+ */
+ZmAccountsPage.prototype._recoveryResetRequest =
+function(response) {
```
Comment (desouzas) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Previously we tabbed `function`, maybe just move them all to the same line as the declaration? Consider this for the rest of these method declarations.

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 9]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+ZmAccountsPage.prototype._recoveryErrorCallback =
+function(response) {
+	var message = response.msg;
+	var code = response.code;
+	var messageBeginning = message.indexOf(": ") + 2; // Simplify the message to description - usually the first colon is the error prefix
+	this.recoveryCodeInputField.setValue("", true);
+	message = message.substring(messageBeginning);
+	// because one particlar exception message is long, check the code and change the message.
+	if(code === "mail.SEND_ABORTED_ADDRESS_FAILURE") {
+		message = ZmMsg.invalidEmailAddress ;
```
Comment (desouzas) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Remove space before `;`

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 10]
```diff
@@ -318,6 +318,381 @@ function(params) {
...
+	var addRecoveryEmailInputContainer = document.getElementById(this._htmlElId + "_RECOVERY_ACCOUNT_TR");
+	var codeRecoveryEmailInputContainer = document.getElementById(this._htmlElId + "_RECOVERY_CODE_TR");
+	var recoveryAccountInput = document.getElementById(this._htmlElId + "_RECOVERY_ACCOUNT");
+	var recoveryCodeInput = document.getElementById(this._htmlElId + "_RECOVERY_CODE");
+
+	// Add constants
+	this.RECOVERY_EMAIL_STATUS = {none: "none", pending: "pending", verified: "verified"}
+
+	// Add critical elements
+	if (addRecoveryEmailInputContainer && recoveryAccountInput && codeRecoveryEmailInputContainer && recoveryCodeInput) {
```
Comment (desouzas) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: We mix styles of: `if (` and `if(` in this file.

File: WebRoot/templates/prefs/Pages.template (hunk 5 above)
Comment (desouzas) [inline WebRoot/templates/prefs/Pages.template]: Commenting out, can we remove?

File: WebRoot/templates/prefs/Pages.template [hunk 11]
```diff
@@ -1087,6 +1087,77 @@ and the container element is replaced with that control.
 							 </td>
 						</tr>
 						<tr><td colspan=2><hr></td></tr>
+					<$ // } $>
```
Comment (desouzas) [inline WebRoot/templates/prefs/Pages.template]: Same as above.

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 12]
```diff
@@ -368,8 +743,8 @@ function(user, sendAs, sendObo, isGrant, refresh) {
 	batchCmd.run();
 };
 
- ZmAccountsPage.prototype._handleDelegateRightsCallback =
- function(user, sendAs, sendObo, isGrant, refresh, result){
+ZmAccountsPage.prototype._handleDelegateRightsCallback =
+function(user, sendAs, sendObo, isGrant, refresh, result){
```
Comment (desouzas) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Does this one need a docblock too?

File: WebRoot/messages/ZmMsg.properties [hunk 13]
```diff
@@ -4510,6 +4510,21 @@ revokeAllTrustedDevicesMsg = Revoking trusted status for all other devices means
...
 # Password Recovery dialog content
+
+recoveryEmailPrefTitle = Password Recovery Account Settings:
+recoveryEmailButtonAdd = Add Recovery Email
+recoveryEmailButtonReset = Reset Recovery Email
+recoveryEmailButtonResend = Resend Code
+recoveryEmailButtonValidate = Verify Code
+recoveryEmailCodeLabel = Enter Code:
+recoveryEmailOptionLabel = Options:
+recoveryEmailMessageAdd = Enter an email address to act your password recovery email on record.
```
Comment (desouzas) [inline WebRoot/messages/ZmMsg.properties]: Grammar error. Use:

> Enter an email address to **set as** your password recovery email.

File: WebRoot/messages/ZmMsg.properties [hunk 14]
```diff
@@ -4510,6 +4510,21 @@ revokeAllTrustedDevicesMsg = Revoking trusted status for all other devices means
...
+recoveryEmailMessagePending = We have sent an email with a validation Code to confirm your recovery email address. Enter the code in the field below and click the associated button to validate this email address. <br><br>Don\u0027t see the email in your Inbox? Check your Junk folder. If you still don't see it, click the \u0027Resend My Code\u0027 button below and we will resend the code to the requested email.
```
Comment (desouzas) [inline WebRoot/messages/ZmMsg.properties]: Use `Validation Code` or `validation code`.

File: WebRoot/messages/ZmMsg.properties [hunk 15]
```diff
@@ -4510,6 +4510,21 @@ revokeAllTrustedDevicesMsg = Revoking trusted status for all other devices means
...
+recoveryEmailMessageVerified = We have validated your recover email address. You can choose to delete this address by clicking on the reset button below.
```
Comment (desouzas) [inline WebRoot/messages/ZmMsg.properties]: Typo:
> We have validated your **recovery** email address.

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js (hunk 8 above)
Comment (naijenn) [reply WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Im adhering to what was in all the other scripts - I won't say I am a fan, but it seems to be the norm here.

Comment (naijenn) [reply WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Im adhering to what was in all the other scripts - I won't say I am a fan, but it seems to be the norm here.

Comment (desouzas) [reply WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: See line 369 for comparison to understand what I mean.

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 16]
```diff
@@ -318,6 +318,376 @@ function(params) {
...
+};
+
+/**
+ * Validate Code Input's Keyup Listener.
+ *
+ * Keyup handler to enable the validate button once the input is considered valid.
+ *
+ */
+ZmAccountsPage.prototype._handleRecoveryCodeInput =
+function(result) {
```
Comment (jdavid-synacor) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Doc block the parameter.

File: WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js [hunk 17]
```diff
@@ -318,6 +318,376 @@ function(params) {
...
+};
+
+/**
+ * Add Email Input's Keyup Listener.
+ *
+ * Keyup handler to enable the add button once the input is considered valid.
+ *
+ */
+ZmAccountsPage.prototype._handleRecoveryEmailInput =
+function(result) {
```
Comment (jdavid-synacor) [inline WebRoot/js/zimbraMail/mail/view/prefs/ZmAccountsPage.js]: Doc block the parameter.

**Pull request merged**